internal/gcsx/random_reader.go

```diff
@@ -329,60 +345,72 @@ func (rr *randomReader) ReadAt(
 		return
 	}
 
-	// Check first if we can read using existing reader. if not, determine which
-	// api to use and call gcs accordingly.
-
-	// When the offset is AFTER the reader position, try to seek forward, within reason.
-	// This happens when the kernel page cache serves some data. It's very common for
-	// concurrent reads, often by only a few 128kB fuse read requests. The aim is to
-	// re-use GCS connection and avoid throwing away already read data.
-	// For parallel sequential reads to a single file, not throwing away the connections
-	// is a 15-20x improvement in throughput: 150-200 MiB/s instead of 10 MiB/s.
-	if rr.reader != nil && rr.start < offset && offset-rr.start < maxReadSize {
-		bytesToSkip := offset - rr.start
-		discardedBytes, copyError := io.CopyN(io.Discard, rr.reader, bytesToSkip)
-		// io.EOF is expected if the reader is shorter than the requested offset to read.
-		if copyError != nil && !errors.Is(copyError, io.EOF) {
-			logger.Warnf("Error while skipping reader bytes: %v", copyError)
+	// Not taking any lock for getting reader type to ensure random read requests do not wait.
+	readInfo := rr.getReadInfo(offset, false)
+	reqReaderType := readerType(readInfo.readType, rr.bucket.BucketType())
+
+	if reqReaderType == RangeReader {
+		rr.mu.Lock()
+		expectedOffset := rr.expectedOffset.Load()
+
+		// Calculating reader type again for zonal buckets in case another read has been served
+		// since last computation. This is to ensure that we don't use range reader incorrectly
+		// when MRD should've been used.
+		if rr.bucket.BucketType().Zonal && readInfo.expectedOffset != expectedOffset {
+			readInfo = rr.getReadInfo(offset, readInfo.seekRecorded)
+			reqReaderType = readerType(readInfo.readType, rr.bucket.BucketType())
 		}
-		rr.start += discardedBytes
-	}
 
-	// If we have an existing reader, but it's positioned at the wrong place,
-	// clean it up and throw it away.
-	// We will also clean up the existing reader if it can't serve the entire request.
-	dataToRead := math.Min(float64(offset+int64(len(p))), float64(rr.object.Size))
-	if rr.reader != nil && (rr.start != offset || int64(dataToRead) > rr.limit) {
-		rr.closeReader()
-		rr.reader = nil
-		rr.cancel = nil
-	}
+		if reqReaderType == MultiRangeReader {
+			rr.mu.Unlock()
```

Comment:
you can add this line also under this if clause:
objectData.Size, err = rr.readFromMultiRangeReader(ctx, p, offset, offset+int64(len(p)), TimeoutForMultiRangeRead)

then extract the rest of the logic into seperate method - readFromRangeReader or something like that.